cpp/src/arrow/record_batch.cc

```diff
@@ -253,14 +253,65 @@ inline void ConvertColumnsToTensor(const RecordBatch& batch, uint8_t* out) {
   using CType = typename arrow::TypeTraits<DataType>::CType;
   auto* out_values = reinterpret_cast<CType*>(out);
 
-  // Loop through all of the columns
-  for (int i = 0; i < batch.num_columns(); ++i) {
-    const auto* in_values = batch.column(i)->data()->GetValues<CType>(1);
-
-    // Copy data of each column
-    memcpy(out_values, in_values, sizeof(CType) * batch.num_rows());
-    out_values += batch.num_rows();
-  }  // End loop through columns
+  if (TypeTraits<DataType>::type_singleton() ==
+      batch.column(0)->type()) {  // If all columns are of same data type
+    // Loop through all of the columns
+    for (int i = 0; i < batch.num_columns(); ++i) {
+      const auto& arr = *batch.column(i);
+      auto data = arr.data();
+      const auto& in_values = data->GetValues<CType>(1);
+
+      // Copy data of each column
+      memcpy(out_values, in_values, sizeof(CType) * batch.num_rows());
+      out_values += batch.num_rows();
+    }  // End loop through columns
+
+  } else {  // If columns have mixed data type
+    // Loop through all of the columns
+    for (int i = 0; i < batch.num_columns(); ++i) {
+      const auto& arr = *batch.column(i);
+      auto data = arr.data();
+
+      // Copy data of each column
+      for (int i = 0; i < arr.length(); ++i) {
+        switch (arr.type_id()) {
+          case Type::UINT8:
+            *out_values++ = static_cast<CType>(data->GetValues<uint8_t>(1)[i]);
```

Comment:
It would be good to check what is the overhead of calling `data->GetValues<uint8_t>(1)` inside the loop, i.e. for each value.

Otherwise you can get those values once before the loop, and then use them similarly as the `out_values`:

```
    const uint8_t* in_values = data->GetValues<uint8_t>(1);
    for (int64_t i = 0; i < arr.length(); ++i) {
      *out_values++ = static_cast<CType>(*in_values++);
    }
```

The fits less nicely in the switch case (moving the loop inside), but so maybe you can avoid the repetition with a macro